pkg/cvo/cvo.go

```diff
@@ -135,119 +174,159 @@ func (optr *Operator) Run(workers int, stopCh <-chan struct{}) {
+		},
 	}
 }
 
-func (optr *Operator) worker() {
-	for optr.processNextWorkItem() {
+func (optr *Operator) worker(queue workqueue.RateLimitingInterface, syncHandler func(string) error) {
+	for processNextWorkItem(queue, syncHandler, optr.syncDegradedStatus) {
 	}
 }
 
-func (optr *Operator) processNextWorkItem() bool {
-	key, quit := optr.queue.Get()
+func processNextWorkItem(queue workqueue.RateLimitingInterface, syncHandler func(string) error, syncDegradedStatus func(err error) error) bool {
+	key, quit := queue.Get()
 	if quit {
 		return false
 	}
-	defer optr.queue.Done(key)
-
-	err := optr.syncHandler(key.(string))
-	optr.handleErr(err, key)
+	defer queue.Done(key)
 
+	err := syncHandler(key.(string))
+	handleErr(queue, err, key, syncDegradedStatus)
 	return true
 }
 
-func (optr *Operator) handleErr(err error, key interface{}) {
+func handleErr(queue workqueue.RateLimitingInterface, err error, key interface{}, syncDegradedStatus func(err error) error) {
 	if err == nil {
-		optr.queue.Forget(key)
+		queue.Forget(key)
 		return
 	}
 
-	if optr.queue.NumRequeues(key) < maxRetries {
+	if queue.NumRequeues(key) < maxRetries {
 		glog.V(2).Infof("Error syncing operator %v: %v", key, err)
-		optr.queue.AddRateLimited(key)
+		queue.AddRateLimited(key)
 		return
 	}
 
-	err = optr.syncDegradedStatus(err)
+	err = syncDegradedStatus(err)
 	utilruntime.HandleError(err)
-	glog.V(2).Infof("Dropping operator %q out of the queue: %v", key, err)
-	optr.queue.Forget(key)
+	glog.V(2).Infof("Dropping operator %q out of the queue %v: %v", key, queue, err)
+	queue.Forget(key)
 }
 
 func (optr *Operator) sync(key string) error {
 	startTime := time.Now()
-	glog.V(4).Infof("Started syncing operator %q (%v)", key, startTime)
+	glog.V(4).Infof("Started syncing cluster version %q (%v)", key, startTime)
 	defer func() {
-		glog.V(4).Infof("Finished syncing operator %q (%v)", key, time.Since(startTime))
+		glog.V(4).Infof("Finished syncing cluster version %q (%v)", key, time.Since(startTime))
 	}()
 
-	// We always run this to make sure CVOConfig can be synced.
-	if err := optr.syncCustomResourceDefinitions(); err != nil {
-		return err
-	}
-
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	config, err := optr.getOrCreateClusterVersion()
 	if err != nil {
 		return err
 	}
 
-	var obj *cvv1.CVOConfig
-	obj, err = optr.cvoConfigLister.CVOConfigs(namespace).Get(name)
-	if apierrors.IsNotFound(err) {
-		obj, err = optr.getConfig()
+	availableCondition := resourcemerge.FindOperatorStatusCondition(config.Status.Conditions, osv1.OperatorAvailable)
+
+	// when we're up to date, limit how frequently we check the payload
+	if config.Status.Generation == config.Generation {
+		if availableCondition != nil && availableCondition.Status == osv1.ConditionTrue && availableCondition.LastTransitionTime.Time.After(time.Now().Add(-optr.minimumUpdateCheckInterval)) {
+			glog.V(2).Infof("Waiting for more time to pass before attempting to resync the payload")
+			return nil
```

Comment:
Seems like you should requeue here.  Are you counting on a resync instead?  I would prefer an explicit requeue.

Reply:
Resync interval at 2m does that for us, yeah.  AvailableUpdates keys off the same thing.  AddAfter isn't enough because need to reconcile every X, not just after the last change.